tests/nnstreamer_rate/unittest_rate.cc

```diff
@@ -0,0 +1,475 @@
+  gst_object_unref (test_data.pipeline);
+}
+
+/**
+ * @brief Test tensor_rate set invalide framerate (negative)
+ */
+TEST (nnstreamer_rate, set_propery_invalid_framerate_n)
+{
+  TestOption option;
+  GstElement *rate;
+  gchar *framerate;
+
+  _set_default_option (option);
+  option.mode = TENSOR_RATE_MODE_PASSTHROUGH;
+
+  ASSERT_TRUE (_setup_pipeline (option));
+
+  rate = gst_bin_get_by_name (GST_BIN (test_data.pipeline), "rate");
+  ASSERT_TRUE (rate != NULL);
+
+  g_object_set (rate, "framerate", "ASDF", NULL);
+  g_object_get (rate, "framerate", &framerate, NULL);
+  EXPECT_STREQ (framerate, DEFAULT_SOURCE_FRAMERATE);
+  g_free (framerate);
+
+  g_object_set (rate, "framerate", "10/0", NULL);
+  g_object_get (rate, "framerate", &framerate, NULL);
+  EXPECT_STREQ (framerate, DEFAULT_SOURCE_FRAMERATE);
+  g_free (framerate);
+
+  gst_object_unref (rate);
+  gst_object_unref (test_data.pipeline);
+}
+
+/**
+ * @brief Test tensor_rate with passthrough mode
+ */
+TEST (nnstreamer_rate, passthrough)
+{
+  TestOption option;
+  GstElement *rate;
+  guint64 in, out, dup, drop;
+
+  _set_default_option (option);
+  option.mode = TENSOR_RATE_MODE_PASSTHROUGH;
+
+  ASSERT_TRUE (_setup_pipeline (option));
+
+  rate = gst_bin_get_by_name (GST_BIN (test_data.pipeline), "rate");
+  ASSERT_TRUE (rate != NULL);
+
+  EXPECT_EQ (setPipelineStateSync (test_data.pipeline, GST_STATE_PLAYING,
+        UNITTEST_STATECHANGE_TIMEOUT), 0);
+
+  g_usleep (G_USEC_PER_SEC / 10);
+
+  g_object_get (rate, "in", &in, NULL);
+  g_object_get (rate, "out", &out, NULL);
+  g_object_get (rate, "duplicate", &dup, NULL);
+  g_object_get (rate, "drop", &drop, NULL);
+
+  EXPECT_EQ (in, option.source_num_buffers);
+  EXPECT_EQ (out, option.source_num_buffers);
+  EXPECT_EQ (dup, 0);
+  EXPECT_EQ (drop, 0);
+
+  EXPECT_EQ (setPipelineStateSync (test_data.pipeline, GST_STATE_NULL,
+        UNITTEST_STATECHANGE_TIMEOUT), 0);
+
+  gst_object_unref (rate);
+  gst_object_unref (test_data.pipeline);
+}
+
+/**
+ * @brief Test tensor_rate with no-throttling mode
+ */
+TEST (nnstreamer_rate, no_throttling)
+{
+  TestOption option;
+  GstElement *rate;
+  guint64 in, out, dup, drop;
+
+  _set_default_option (option);
+  option.mode = TENSOR_RATE_MODE_NO_THROTTLE;
+  option.target_framerate = g_strdup ("15/1");
+
+  ASSERT_TRUE (_setup_pipeline (option));
+
+  rate = gst_bin_get_by_name (GST_BIN (test_data.pipeline), "rate");
+  ASSERT_TRUE (rate != NULL);
+
+  EXPECT_EQ (setPipelineStateSync (test_data.pipeline, GST_STATE_PLAYING,
+        UNITTEST_STATECHANGE_TIMEOUT), 0);
+
+  g_usleep (G_USEC_PER_SEC / 10);
+
+  g_object_get (rate, "in", &in, NULL);
+  g_object_get (rate, "out", &out, NULL);
+  g_object_get (rate, "duplicate", &dup, NULL);
+  g_object_get (rate, "drop", &drop, NULL);
+
+  EXPECT_EQ (in, option.source_num_buffers);
+  EXPECT_EQ (dup, 0);
+
+  /** we don't expect the exact values */
+  EXPECT_GE (out, (guint64) ((option.source_num_buffers / 2) * 0.95));
```

Comment:
could you please run these tests on gbs armv7l

Reply:
Right.. I think it could have some timing issues.

Reply:
I increased the sleep time to 1 sec, and checked all tests works well.

Reply:
Giving a long sleep time is not a good solution.
How about using `wait_pipeline_process_buffers`.?

Reply:
It seems that `wait_pipeline_process_buffers` is not a generic API. If you're saying the internal function in `unittest_sink`,
I am not sure the same mechanism works as it may control number of output buffers,

Reply:
Right, it is impossible to use `wait_pipeline_process_buffers` in this test.
However, I think it seems better to ensure the integrity of the test through check and repeat.
A second doesn't mean much.

Reply:
@gichan-jang OK, I understand your concerns. I updated codes to wait the eos event instead of the constant timeout.